src/DotNetty.Codecs.Redis/RedisDecoder.cs

```diff
@@ -0,0 +1,389 @@
+﻿
+namespace DotNetty.Codecs.Redis
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics.Contracts;
+    using System.Text;
+    using DotNetty.Buffers;
+    using DotNetty.Codecs.Redis.Messages;
+    using DotNetty.Common.Utilities;
+    using DotNetty.Transport.Channels;
+
+    public sealed class RedisDecoder : ByteToMessageDecoder
+    {
+        readonly IRedisMessagePool messagePool;
+        readonly int maximumInlineMessageLength;
+        readonly ToPositiveLongProcessor toPositiveLongProcessor = new ToPositiveLongProcessor();
+
+        enum State
+        {
+            DecodeType,
+            DecodeInline, // SIMPLE_STRING, ERROR, INTEGER
+            DecodeLength, // BULK_STRING, ARRAY_HEADER
+            DecodeBulkStringEndOfLine,
+            DecodeBulkStringContent,
+        }
+
+        // current decoding states
+        State state = State.DecodeType;
+        RedisMessageType type;
+        int remainingBulkLength;
+
+        public RedisDecoder()
+            : this(RedisConstants.MaximumInlineMessageLength, FixedRedisMessagePool.Default)
+        { }
+
+        internal RedisDecoder(int maximumInlineMessageLength, IRedisMessagePool messagePool)
```

Comment:
in turn, why is interface internal? it makes sense to open both ctor and intf to allow to implement use case specific pool.